uefi/src/memory_descriptor.rs

```diff
@@ -0,0 +1,45 @@
+use bootloader_api::info::MemoryRegionKind;
+use bootloader_x86_64_common::legacy_memory_region::LegacyMemoryRegion;
+use uefi::table::boot::{MemoryDescriptor, MemoryType};
+use x86_64::PhysAddr;
+
+#[derive(Debug, Copy, Clone)]
+pub struct UefiMemoryDescriptor(pub MemoryDescriptor);
+
+const PAGE_SIZE: u64 = 4096;
+pub const KERNEL_MEMORY_TYPE: MemoryType = MemoryType::custom(0x80000000);
+
+impl<'a> LegacyMemoryRegion for UefiMemoryDescriptor {
+    fn start(&self) -> PhysAddr {
+        PhysAddr::new(self.0.phys_start)
+    }
+
+    fn len(&self) -> u64 {
+        self.0.page_count * PAGE_SIZE
+    }
+
+    fn kind(&self) -> MemoryRegionKind {
+        match self.0.ty {
+            MemoryType::CONVENTIONAL => MemoryRegionKind::Usable,
+            other if other == KERNEL_MEMORY_TYPE => MemoryRegionKind::Kernel,
+            other => MemoryRegionKind::UnknownUefi(other.0),
+        }
+    }
+
+    fn usable_after_bootloader_exit(&self) -> bool {
+        match self.0.ty {
+            MemoryType::CONVENTIONAL => true,
+            MemoryType::LOADER_CODE
+            | MemoryType::LOADER_DATA
+            | MemoryType::BOOT_SERVICES_CODE
+            | MemoryType::BOOT_SERVICES_DATA
+            | MemoryType::RUNTIME_SERVICES_CODE
+            | MemoryType::RUNTIME_SERVICES_DATA => {
```

Comment:
According to the UEFI spec, regions of type `RUNTIME_SERVICES_CODE` and `RUNTIME_SERVICES_DATA` should be preserved by the OS loader and the OS.

Reply:
Also if you don't use any UEFI runtime services anymore?

Reply:
I think so, yes.

> The memory in this range is to be preserved by the UEFI OS loader and OS
in the working and ACPI S1-S3 states. 
>
> Source: UEFI spec 2.10 Table 7.10 Memory Type Usage after ExitBootServices()

Reply:
Ok, I changed the code to preserve these sections in 667e57f.